tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/maildelivery/impl/MailDeliveryService.kt

```diff
@@ -126,55 +129,67 @@ private constructor(
-            )
+        val globalVars = unwrapStruct(globalVarsStruct)
 
-            mqSender.queueJob(jobMessageFactory.createSendEmailJob(
-                    requestedSendTime = requestedSendTime,
-                    sentMailMessageBodyId = mailMessageBody.id,
-                    emailIds = listOf(emailDBO.id)
-            ))
-            emailDAO.save(emailDBO)
-
-            responseObserver.onNext(
-                    EmailQueueStatus.newBuilder()
-                            .setEmailId(emailDBO.id.toGrpcInternal())
-                            .setQueued(Queued.getDefaultInstance())
-                            .setTransactionId(transaction.id.toGrpcInternal())
-                            .setRecipient(emailDBO.recipient.toGrpcInternal())
-                            .build()
-            )
+        ebeanServer.createTransaction().use {
+            for (recipient in request.recipientsList) {
+                // TODO Check if recipient is blocked
+                val recipientVars = unwrapStruct(recipient.templateVars)
+                val htmlBody = mergeTemplate(template.htmlTemplate, globalVars, recipientVars)
+                val plaintextBody = mergeTemplate(template.plaintextTemplate, globalVars, recipientVars)
+                val mailMessageBody = SentMailMessageBodyDBO(
+                        wrapAndPackBody(
+                                request = request,
+                                htmlBody = htmlBody.emptyToNull(),
+                                plaintextBody = plaintextBody.emptyToNull(),
+                                subject = mergeTemplate(template.subject, globalVars, recipientVars)
+                        )
+                )
+                val emailDBO = EmailDBO(
+                        recipient = recipient.toNamedEmail(),
+                        transaction = transaction,
+                        sentMailMessageBody = mailMessageBody
+                )
+                emailDAO.save(emailDBO)
+
+                mqSender.queueJob(jobMessageFactory.createSendEmailJob(
+                        requestedSendTime = requestedSendTime,
+                        sentMailMessageBodyId = mailMessageBody.id,
+                        emailIds = listOf(emailDBO.id)
+                ))
+
+                responseObserver.onNext(
+                        EmailQueueStatus.newBuilder()
+                                .setEmailId(emailDBO.id.toGrpcInternal())
+                                .setQueued(Queued.getDefaultInstance())
+                                .setTransactionId(transaction.id.toGrpcInternal())
+                                .setRecipient(emailDBO.recipient.toGrpcInternal())
+                                .build()
+                )
+            }
+            responseObserver.onCompleted()
         }
-        responseObserver.onCompleted()
+    }
+
+    private fun unwrapStruct(struct: Struct): HashMap<String, Any> {
+        val objectMapper = ObjectMapper().registerKotlinModule()
```

Comment:
Please don't create a new one every time. dbObjectMapper is there, please use it or create a new one in a companion object